File: src/metreos/util/PerfTest.java
package metreos.util;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Framework for running performance tests.
 */
abstract public class PerfTest
{
	/**
	 * Constructs the PerfTest.
	 * @param descr the description of the test.
	 * @param runtime the number of seconds to run each test cycle.
	 * @param count the number of test cycles to run.
	 */
	public PerfTest( String descr, int runtime, int count )
	{
		this.descr = descr;
		this.runtime = runtime;
		this.count = count;
	}
	
	private final String descr;
	
	private final int runtime;
	
	private final int count;
	
	/**
	 * Runs the perf test.
	 * @throws Exception 
	 */
	final public void run() throws Exception
    {
		long oldn = 0;
		long n = 1;
		double t = 0;
		while (t < 1)
		{
			if (t > 0.1)
				System.out.printf( "%s: %d took %f, trying %d to get >= 1 second\n", descr, oldn, t, n );
			oldn = n;
			t = runOne( n );
			n *= 2;
		}
		n = oldn;
		
		int k = 2;
		n = (long) (k * n / t);
		System.out.printf( "%s: %d took %f, trying %d for %d seconds\n", descr, oldn, t, n, k );
		oldn = n;
		t = runOne( n );
		
		k = 4;
		n = (long) (k * n / t);
		System.out.printf( "%s: %d took %f, trying %d for %d seconds\n", descr, oldn, t, n, k );
		oldn = n;
		t = runOne( n );
		
		n = (long) (runtime * n / t);
		System.out.printf( "%s: %d took %f, using %d for %d seconds\n",
			descr, oldn, t, n, runtime );
		
		List<Double> list = new ArrayList<Double>( count );
		double sum = 0;
		for (int i = 1; i <= count; i++)
		{
			t = runOne( n );
			double r = n/t;
			list.add( r );
			sum += r;
			System.out.printf( "%s %d/%d\t%d\t%f\t%f\n",
				descr, i, count, n, t, r );
		}
		Collections.sort( list );
		System.out.printf( "%s min\t%f\n", descr, list.get( 0 ) );
		System.out.printf( "%s median\t%f\n", descr, list.get( count/2 ) );
		System.out.printf( "%s avg\t%f\n", descr, sum/count );
		System.out.printf( "%s max\t%f\n", descr, list.get( count-1 ) );
    }
	
	final private double runOne( long n ) throws Exception
	{
	    long t0 = System.nanoTime();
	    run( n );
	    long t1 = System.nanoTime();
	    return (t1 - t0) / 1000000000.0;
	}
	
	/**
	 * Runs the test with the specific number of times.
	 * @param n
	 * @throws Exception 
	 */
	abstract public void run( long n ) throws Exception;
}

File: src/etch/bindings/csharp/util/PerfTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using Etch.Transport;

namespace Etch.Util
{
    /// <summary>
    /// Framework for running performance tests.
    /// </summary>
    public class PerfTest
    {
        private readonly String _descr;
        private readonly int _runtime;
        private readonly int _count;
        private readonly RunTimes _r;

        /// <summary>
        /// Construct the perftest
        /// </summary>
        /// <param name="descr">description of the test</param>
        /// <param name="runtime">the number of seconds to run each test cycle</param>
        /// <param name="count">the number of test cycles to run</param>
        public PerfTest( String descr, int runtime, int count, RunTimes rt )
        {
            _descr = descr;
            _runtime = runtime;
            _count = count;
            _r = rt;
        }

        /// <summary>
        /// Runs the PerfTest
        /// </summary>
        public void Run( )
        {
            
            long oldn = 0;
            long n = 1;
            double t = 0;

            while (t < 1)
		    {
			    if (t > 0.1)
				    Console.WriteLine( "{0}: {1} took {2}, trying {3} to get >= 1 second\n", _descr, oldn, t, n );
			    oldn = n;
			    t = RunOne( n );
			    n *= 2;
		    }
		    n = oldn;

            int k = 2;
		    n = (long) (k * n / t);
		    Console.WriteLine( "{0}: {1} took {2}, trying {3} for {4} seconds\n", _descr, oldn, t, n, k );
		    oldn = n;
		    t = RunOne( n );

            k = 4;
		    n = (long) (k * n / t);
		    Console.WriteLine( "{0}: {1} took {2}, trying {3} for {4} seconds\n", _descr, oldn, t, n, k );
		    oldn = n;
		    t = RunOne( n );
    		
		    n = (long) (_runtime * n / t);
            Console.WriteLine( "{0}: {1} took {2}, using {3} for {4} seconds\n",
			    _descr, oldn, t, n, _runtime );

            List<double> list = new List<double>( _count );
            double sum = 0;
		    for (int i = 1; i <= _count; i++)
		    {
			    t = RunOne( n );
			    double r = n/t;
			    list.Add( r );
			    sum += r;
			    Console.WriteLine( "{0} {1}/{2}\t{3}\t{4}\t{5}\n",
				    _descr, i, _count, n, t, r );
		    }

            list.Sort();
            Console.WriteLine( "{0} min\t{1}\n", _descr, list[ 0 ] );
            Console.WriteLine( "{0} median\t{1}\n", _descr, list[ _count/2 ] );
		    Console.WriteLine( "{0} avg\t{1}\n", _descr, sum/_count );
		    Console.WriteLine( "{0} max\t{1}\n", _descr, list[ _count-1 ] );
        }

        private double RunOne( long n )
        {
            long t0 = Etch.Util.HPTimer.Now();
            _r( n );
            long t1 = Etch.Util.HPTimer.NsSince( t0 );
            return ( t1 )/ 1000000000.0;
        }

        /// <summary>
        /// Runs the test with the specific number of times.
        /// </summary>
        /// <param name="n"></param>
        public delegate void RunTimes( long n );
    }
}
